### PR TITLE
Initial version of properties analyzer

### LINKED ISSUE #9883: [Built-in analyzer] Undefined property is used
### Background

This issue tracks one of the BuildCheck analyzers we would like to ship in-box with MSBuild.

### Goal

Implement an analyzer with the following rule: No property is used before it's defined, except for conditions.

### Notes

A non-existent property is silently expanded to an empty string, which can lead to unexpected behavior. We would like to flag this as a violation, except for cases where the property is tested in a condition like so `'$(MyProp)' == ''`.

### Implementation Update

Fully working prototype: https://github.com/dotnet/msbuild/pull/10009
For easier reviewing and less conflucts during other ongoing work, the payload will be delivered in 3 subsequent payloads:
* [x] Expander refactoring https://github.com/dotnet/msbuild/pull/10102
* [ ] Wiring the data https://github.com/dotnet/msbuild/pull/10237
* [ ] Actuall BuildCheck implementation - this is **blocked by**:
   * [x] https://github.com/dotnet/msbuild/issues/10123
   * [ ] https://github.com/orgs/dotnet/projects/373/views/7?pane=issue&itemId=57851137

### LINKED ISSUE #10427: [Bug] [BuildCheck] [Documentation]: Default severity of build-in analyzer rules should be documented.
We should add default severity of the build-in analyzer rules to https://github.com/dotnet/msbuild/blob/main/documentation/specs/BuildCheck/Codes.md

### PR DESCRIPTION
**Pull request by JanKrivanek**
Fixes #9883
Fixes #10427

### Context

This is a proposal of multirule integrated internal build check that checks for:
* Property being used without being initialized
* Property being used before being initialized
* Property being initialized but never used

### Testing
 Targetted unittests

### UX

sample output from test for current proposal:

```
MSBUILD : warning : <path>\myProj.proj (24,10): BC0202: Property: [MyPropT2] first declared/initialized at [<path>\myProj.proj (26,1)] used before it was initialized.
MSBUILD : warning : <path>\myProj.proj (13,3): BC0203: Property: [MyProp13] was declared/initialized, but it was never used.
MSBUILD : warning : <path>\myProj.proj (13,3): BC0201: Property: [MyProp11] was accessed, but it was never initialized.

```

### CODE REVIEW

File: documentation/specs/BuildCheck/Codes.md [hunk 1]
```diff
@@ -43,6 +45,48 @@ Relying on environment variables introduces variability and unpredictability, as
 
 This practice can result in inconsistent build outcomes and makes debugging difficult, since environment variables are external to project files and build scripts. To ensure consistent and reproducible builds, avoid using environment variables. Instead, explicitly pass properties using the /p option, which offers better control and traceability.
 
+## <a name="BC0201"></a>BC0201 - Usage of undefined property.
+
+"A property that is accessed should be declared first."
+
+This check indicates that a property was acessed without being declared (the declaration might have happen later - see [BC0202](#BC0202) for such checking). Only accessing in the configured scope (by default it's the project file only) are checked.
```
Comment (MichalPavlik) [inline documentation/specs/BuildCheck/Codes.md:52]: nit: acessed -> accessed

File: documentation/specs/BuildCheck/Codes.md [hunk 2]
```diff
@@ -43,6 +45,48 @@ Relying on environment variables introduces variability and unpredictability, as
...
+
+There are couple cases which are allowed by the check:
+
+* Selfreferencing declaration is allowed - e.g.:
+  `<ChainProp>$(ChainProp)</ChainProp>`
+
+* Checking the property for emptyness - e.g.:
+  `<PropertyGroup Condition="'$(PropertyThatMightNotBeDefined)' == ''">`
+
+* Any usage of property in condition. This can be opted out vie the configuration `AllowUninitializedPropertiesInConditions` - e.g.:
```
Comment (MichalPavlik) [inline documentation/specs/BuildCheck/Codes.md:62]: nit: vie -> via

File: documentation/specs/BuildCheck/Codes.md [hunk 3]
```diff
@@ -43,6 +45,48 @@ Relying on environment variables introduces variability and unpredictability, as
...
+  build_check.BC0202.AllowUninitializedPropertiesInConditions=false
+  ```
+
+  BC0201 and BC0202 must have same value for the optional switch - as both operate on top of same data and same filtering.
+
+## <a name="BC0202"></a>BC0202 - Property first declared after it was used.
+
+"A property should be declared before it is first used."
+
+This check indicates that a property was acessed before it was declared. The default scope of this rule is the project file only. The scope captures the read and write operations as well. So this rule reports:
```
Comment (MichalPavlik) [inline documentation/specs/BuildCheck/Codes.md:76]: nit: acessed -> accessed

File: src/Build/BuildCheck/Checks/PropertiesUsageCheck.cs [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+using Microsoft.Build.Experimental.BuildCheck;
+using Microsoft.Build.Experimental.BuildCheck.Infrastructure;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Experimental.BuildCheck.Checks;
+
+internal class PropertiesUsageCheck : InternalCheck
+{
+    private static readonly CheckRule _usedBeforeInitializedRule = new CheckRule("BC0201", "PropertyUsedBeforeDeclared",
+        "A property that is accessed should be declared first.",
```
Comment (MichalPavlik) [inline src/Build/BuildCheck/Checks/PropertiesUsageCheck.cs:20]: Is this text hardcoded by design and we will not localize?

**Review by MichalPavlik — APPROVED** [review]

File: src/Build/BuildCheck/Checks/PropertiesUsageCheck.cs (hunk 4 above)
Comment (JanKrivanek) [reply src/Build/BuildCheck/Checks/PropertiesUsageCheck.cs:20]: Yeah - localization of Checks is not part of NET9. We'll need to design and implement it with next version

**Review by surayya-MS — APPROVED** [review]
Looks good to me! made some suggestions but not critical

File: src/Build/BackEnd/Components/Logging/EvaluationLoggingContext.cs [hunk 5]
```diff
@@ -30,6 +31,8 @@ public EvaluationLoggingContext(ILoggingService loggingService, BuildEventContex
         public void LogProjectEvaluationStarted()
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/BackEnd/Components/Logging/EvaluationLoggingContext.cs:31]: Renamed to ProcessProjectEvaluationStarted (to be consistent with naming in BuildEngineDataRouter)

**Review by Nirmal4G — COMMENTED** [review]
Forgot to Submit the review.

File: src/Shared/IElementLocation.cs [hunk 6]
```diff
@@ -20,7 +20,7 @@ internal interface IElementLocation : IMsBuildElementLocation, ITranslatable { }
     /// This is currently internal - but it is prepared to be made public once it will be needed by other public BuildCheck OM
     /// (e.g. by property read/write OM)
     /// </remarks>
-    internal interface IMsBuildElementLocation
+    public interface IMsBuildElementLocation
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Shared/IElementLocation.cs:23]: Any review at any time is appreciated ;-)

**Pull request merged**